File: backend/app/units.py
"""Unit handling for numeric questions: what a value means, and how to show it in another.

A question about temperature or a measurement is meaningless without its unit, which is
the bug that put "the temperature question had no unit" on the list. An author sets a
``unit`` and, optionally, a ``display_unit`` to also show (e.g. a freezer logged in
Celsius, read by someone who thinks in Fahrenheit). The converter turns one into the
other so the respondent sees both live, and the results keep the author's unit.

Temperature, length, mass and volume are pure formulas, so conversion is exact and needs
no network. Currency is deliberately absent: an exchange rate is not a formula, it is a
feed, and wiring one in is its own decision (source, caching, a failure mode). When that
lands, ``_UNITS`` gains a ``currency`` dimension and ``convert`` needs a rate table;
nothing else here changes.

Each unit is (dimension, factor_to_base, offset_to_base) where
``base = value * factor + offset``. Linear units have offset 0; temperature does not.
``K``/metre/kilogram/litre are the bases.
"""

from dataclasses import dataclass

# dimension -> (factor to base, offset to base). base value = value * factor + offset.
_UNITS: dict[str, tuple[str, float, float]] = {
    # Temperature, base Kelvin.
    "K": ("temperature", 1.0, 0.0),
    "C": ("temperature", 1.0, 273.15),
    "F": ("temperature", 5.0 / 9.0, 273.15 - 32.0 * 5.0 / 9.0),
    # Length, base metre.
    "m": ("length", 1.0, 0.0),
    "cm": ("length", 0.01, 0.0),
    "mm": ("length", 0.001, 0.0),
    "km": ("length", 1000.0, 0.0),
    "in": ("length", 0.0254, 0.0),
    "ft": ("length", 0.3048, 0.0),
    "mi": ("length", 1609.344, 0.0),
    # Mass, base kilogram.
    "kg": ("mass", 1.0, 0.0),
    "g": ("mass", 0.001, 0.0),
    "mg": ("mass", 1e-6, 0.0),
    "t": ("mass", 1000.0, 0.0),
    "lb": ("mass", 0.45359237, 0.0),
    "oz": ("mass", 0.028349523125, 0.0),
    # Volume, base litre.
    "l": ("volume", 1.0, 0.0),
    "ml": ("volume", 0.001, 0.0),
    "gal": ("volume", 3.785411784, 0.0),
}

DIMENSION_LABELS: dict[str, str] = {
    "temperature": "Temperature",
    "length": "Length",
    "mass": "Mass",
    "volume": "Volume",
}

# How a unit reads when shown to a human. Most are the bare symbol; temperature gets the
# degree sign so "32 C" is not mistaken for a count.
UNIT_LABELS: dict[str, str] = {
    "C": "°C",
    "F": "°F",
    "K": "K",
    "m": "m",
    "cm": "cm",
    "mm": "mm",
    "km": "km",
    "in": "in",
    "ft": "ft",
    "mi": "mi",
    "kg": "kg",
    "g": "g",
    "mg": "mg",
    "t": "t",
    "lb": "lb",
    "oz": "oz",
    "l": "L",
    "ml": "mL",
    "gal": "gal",
}


def unit_symbol(unit: str) -> str:
    return UNIT_LABELS.get(unit, unit)


def is_known_unit(unit: str | None) -> bool:
    return unit in _UNITS


def unit_dimension(unit: str) -> str | None:
    entry = _UNITS.get(unit)
    return entry[0] if entry else None


def can_convert(frm: str | None, to: str | None) -> bool:
    """Both units known, same dimension, and not identical (no point showing 5 m = 5 m)."""
    if not frm or not to or frm == to:
        return False
    dim_f = unit_dimension(frm)
    return dim_f is not None and dim_f == unit_dimension(to)


def convert(value: float, frm: str, to: str) -> float:
    """Convert ``value`` from ``frm`` to ``to``. Raises ValueError if they are not
    convertible (unknown unit, or different dimensions)."""
    f = _UNITS.get(frm)
    t = _UNITS.get(to)
    if f is None or t is None or f[0] != t[0]:
        raise ValueError(f"cannot convert {frm!r} to {to!r}")
    base = value * f[1] + f[2]
    return (base - t[2]) / t[1]


def _format(value: float) -> str:
    """Compact, human number: integers stay integers, others get a few significant figures."""
    if value == int(value):
        return str(int(value))
    return f"{value:.4g}"


def conversion_label(value: float, frm: str, to: str) -> str | None:
    """``"32 °C = 89.6 °F"`` for a convertible pair, or None when there is nothing to show."""
    if not can_convert(frm, to):
        return None
    return (
        f"{_format(value)} {unit_symbol(frm)} = "
        f"{_format(convert(value, frm, to))} {unit_symbol(to)}"
    )


@dataclass(frozen=True)
class UnitChoice:
    value: str
    label: str
    dimension: str


def unit_choices() -> list[UnitChoice]:
    """Every unit, for a grouped dropdown, in a stable dimension/label order."""
    order = ["temperature", "length", "mass", "volume"]
    choices = [UnitChoice(value=u, label=u, dimension=dim) for u, (dim, _, _) in _UNITS.items()]
    choices.sort(key=lambda c: (order.index(c.dimension), c.label))
    return choices

File: frontend/lib/units.ts
// Client-side mirror of backend/app/units.py. Pure formula, no network, so the
// respondent sees a live conversion as they type. Keep the factors/offsets in step with
// the backend: the values stored are plain numbers, but the labels the author and
// respondent read must agree. Currency is intentionally absent (rates need a feed).

export type Dimension = "temperature" | "length" | "mass" | "volume";

type UnitDef = { dimension: Dimension; factor: number; offset: number };

const UNITS: Record<string, UnitDef> = {
  // Temperature, base Kelvin.
  K: { dimension: "temperature", factor: 1, offset: 0 },
  C: { dimension: "temperature", factor: 1, offset: 273.15 },
  F: { dimension: "temperature", factor: 5 / 9, offset: 273.15 - (32 * 5) / 9 },
  // Length, base metre.
  m: { dimension: "length", factor: 1, offset: 0 },
  cm: { dimension: "length", factor: 0.01, offset: 0 },
  mm: { dimension: "length", factor: 0.001, offset: 0 },
  km: { dimension: "length", factor: 1000, offset: 0 },
  in: { dimension: "length", factor: 0.0254, offset: 0 },
  ft: { dimension: "length", factor: 0.3048, offset: 0 },
  mi: { dimension: "length", factor: 1609.344, offset: 0 },
  // Mass, base kilogram.
  kg: { dimension: "mass", factor: 1, offset: 0 },
  g: { dimension: "mass", factor: 0.001, offset: 0 },
  mg: { dimension: "mass", factor: 1e-6, offset: 0 },
  t: { dimension: "mass", factor: 1000, offset: 0 },
  lb: { dimension: "mass", factor: 0.45359237, offset: 0 },
  oz: { dimension: "mass", factor: 0.028349523125, offset: 0 },
  // Volume, base litre.
  l: { dimension: "volume", factor: 1, offset: 0 },
  ml: { dimension: "volume", factor: 0.001, offset: 0 },
  gal: { dimension: "volume", factor: 3.785411784, offset: 0 },
};

const UNIT_SYMBOLS: Record<string, string> = {
  C: "°C",
  F: "°F",
  K: "K",
  m: "m",
  cm: "cm",
  mm: "mm",
  km: "km",
  in: "in",
  ft: "ft",
  mi: "mi",
  kg: "kg",
  g: "g",
  mg: "mg",
  t: "t",
  lb: "lb",
  oz: "oz",
  l: "L",
  ml: "mL",
  gal: "gal",
};

const DIMENSION_ORDER: Dimension[] = ["temperature", "length", "mass", "volume"];

export function isKnownUnit(unit: string | null | undefined): boolean {
  return !!unit && unit in UNITS;
}

export function unitDimension(unit: string | null | undefined): Dimension | null {
  const def = unit ? UNITS[unit] : undefined;
  return def ? def.dimension : null;
}

export function canConvert(from: string | null | undefined, to: string | null | undefined): boolean {
  if (!from || !to || from === to) return false;
  const a = UNITS[from];
  const b = UNITS[to];
  return !!a && !!b && a.dimension === b.dimension;
}

export function convert(value: number, from: string, to: string): number {
  const f = UNITS[from];
  const t = UNITS[to];
  if (!f || !t || f.dimension !== t.dimension) {
    throw new Error(`cannot convert ${from} to ${to}`);
  }
  const base = value * f.factor + f.offset;
  return (base - t.offset) / t.factor;
}

function formatValue(value: number): string {
  if (Number.isInteger(value)) return String(value);
  return value.toPrecision(4).replace(/\.?0+$/, "");
}

export function unitSymbol(unit: string | null | undefined): string {
  if (!unit) return "";
  return UNIT_SYMBOLS[unit] ?? unit;
}

export function conversionLabel(value: number, from: string, to: string): string | null {
  if (!canConvert(from, to)) return null;
  return `${formatValue(value)} ${unitSymbol(from)} = ${formatValue(convert(value, from, to))} ${unitSymbol(to)}`;
}

export interface UnitChoice {
  value: string;
  label: string;
  dimension: Dimension;
}

export function unitChoices(): UnitChoice[] {
  const choices: UnitChoice[] = Object.entries(UNITS).map(([value, def]) => ({
    value,
    label: unitSymbol(value),
    dimension: def.dimension,
  }));
  choices.sort(
    (a, b) =>
      DIMENSION_ORDER.indexOf(a.dimension) - DIMENSION_ORDER.indexOf(b.dimension) ||
      a.label.localeCompare(b.label),
  );
  return choices;
}

export function unitsByDimension(): Record<Dimension, UnitChoice[]> {
  const grouped = {} as Record<Dimension, UnitChoice[]>;
  for (const dim of DIMENSION_ORDER) grouped[dim] = [];
  for (const choice of unitChoices()) grouped[choice.dimension].push(choice);
  return grouped;
}
